src/sync/sync_session.cpp

```diff
@@ -259,23 +259,39 @@ const SyncSession::State& SyncSession::State::active = Active();
 const SyncSession::State& SyncSession::State::dying = Dying();
 const SyncSession::State& SyncSession::State::inactive = Inactive();
 
-SyncSession::SyncSession(SyncClient& client, std::string realm_path, SyncConfig config, bool force_client_resync)
-: m_handle_refresh([this](util::Optional<app::AppError> error) {
-    if (error) {
-        // 10 seconds is arbitrary
-        std::this_thread::sleep_for(std::chrono::milliseconds(10000));
-        if (m_state == &State::active) {
-            user()->refresh_custom_data(m_handle_refresh);
+std::function<void(util::Optional<app::AppError>)> _impl::handle_refresh(std::shared_ptr<SyncSession> session) {
+    return [session](util::Optional<app::AppError> error) {
+        if (!session)
+            return;
+
+        auto session_user = session->user();
+        auto is_user_expired = session_user &&
+            session_user->refresh_jwt().expires_at < std::chrono::duration_cast<std::chrono::seconds>(std::chrono::system_clock::now().time_since_epoch()).count();
+        
+        if (is_user_expired) {
+            std::unique_lock<std::mutex> lock(session->m_state_mutex);
+            session->cancel_pending_waits(lock, error ? error->error_code : std::error_code());
+            if (session->m_config.error_handler) {
+                session->m_config.error_handler(session,
+                                                SyncError(SyncError::ProtocolError::bad_authentication, "expired refresh token", true));
+            }
+        } else if (error || !session_user) {
+            // 10 seconds is arbitrary
+            std::this_thread::sleep_for(std::chrono::milliseconds(10000));
```

Comment:
What is this sleep trying to do?

Reply:
I was mimicking what the old cocoa refresh did– giving 10 seconds between retrying an access token refresh.